README.md
# Flutter Web Frame

[![Pub](https://img.shields.io/pub/v/flutter_web_frame.svg)](https://pub.dev/packages/flutter_web_frame)


Make the frame max width / size when on large devices such as Web or Desktop.

This is very suitable to be used to limit the size of content, if your application is not yet available a responsive version if it is run on multiple devices / platforms.

## Preview

![Preview](https://i.ibb.co/hR9j1RJ/preview.png)

## Install

Follow the instructions to install it [here](https://pub.dev/packages/flutter_web_frame)


## Example Usage

```dart
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:flutter_web_frame/flutter_web_frame.dart';

void main() {
  runApp(MyAp());
}

class MyAp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return FlutterWebFrame(
      builder: (context) {
        return MaterialApp(
          home: Scaffold(
            appBar: AppBar(
              title: Text('Flutter Web Frame'),
            ),
            body: Center(
              child: Text('Body Text'),
            ),
          ),
        );
      },
      maximumSize: Size(475.0, 812.0), // Maximum size
      enabled: kIsWeb, // default is enable, when disable content is full size
      backgroundColor: Colors.grey, // Background color/white space
    );
  }
}
```

CHANGELOG.md
## [1.0.0] - 06/07/2023

* Migrate to latest flutter
* Enhance example app

## [0.0.3] - 18/05/2022

* Migrate to support Flutter 3
* Add clip behavior

## [0.0.2] - 08/04/2021

* Fix typo

## [0.0.1] - 08/04/2021

* Initial release.

lib/flutter_web_frame.dart
library flutter_web_frame;

import 'package:flutter/material.dart';

import 'src/frame_content.dart';
import 'src/media_query_observer.dart';

class FlutterWebFrame extends StatefulWidget {
  /// If not [enabled], the [child] is used directly.
  final bool enabled;

  /// The previewed widget.
  ///
  /// It is common to give the root application widget.
  final WidgetBuilder builder;

  /// Background color in white space
  final Color? backgroundColor;

  /// Maximum size
  final Size maximumSize;

  /// Clip behavior
  final Clip clipBehavior;

  const FlutterWebFrame({
    Key? key,
    required this.builder,
    this.enabled = true,
    this.backgroundColor,
    required this.maximumSize,
    this.clipBehavior = Clip.none,
  }) : super(key: key);

  /// A global builder that should be inserted into [WidgetApp]'s builder
  /// to simulated the simulated device screen and platform properties.
  static Widget appBuilder(
    BuildContext context,
    Widget? widget,
  ) {
    if (!_isEnabled(context)) {
      return widget ?? SizedBox();
    }

    if (!_isEnabled(context)) return widget ?? SizedBox();

    return MediaQuery(
      data: _mediaQuery(context),
      child: Theme(
        data: Theme.of(context).copyWith(
          visualDensity: _isEnabled(context) ? VisualDensity.standard : null,
        ),
        child: widget ?? SizedBox(),
      ),
    );
  }

  static MediaQueryData _mediaQuery(BuildContext context) {
    return FrameContent.mediaQuery(
      context,
      context
              .findAncestorStateOfType<_FlutterWebFrameState>()
              ?.widget
              .maximumSize ??
          Size(375.0, 812.0),
    );
  }

  static bool _isEnabled(BuildContext context) {
    final state = context.findAncestorStateOfType<_FlutterWebFrameState>();
    return state != null && state.widget.enabled;
  }

  @override
  _FlutterWebFrameState createState() => _FlutterWebFrameState();
}

class _FlutterWebFrameState extends State<FlutterWebFrame> {
  @override
  Widget build(BuildContext context) {
    if (!widget.enabled) {
      return Builder(
        builder: widget.builder,
      );
    }

    return Container(
      color: widget.backgroundColor ?? Theme.of(context).dividerColor,
      child: AnimatedSwitcher(
        duration: const Duration(milliseconds: 400),
        child: MediaQueryObserver(
          child: Builder(
            builder: (context) => Stack(
              fit: StackFit.loose,
              alignment: Alignment.center,
              children: [
                Positioned.fill(
                  left: 0,
                  right: 0,
                  top: 0,
                  bottom: 0,
                  child: widget.enabled
                      ? Builder(
                          builder: _buildPreview,
                        )
                      : Builder(
                          builder: widget.builder,
                        ),
                ),
              ],
            ),
          ),
        ),
      ),
    );
  }

  Widget _buildPreview(BuildContext context) {
    return FittedBox(
      fit: BoxFit.contain,
      child: RepaintBoundary(
        child: FrameContent(
          size: widget.maximumSize,
          clipBehavior: widget.clipBehavior,
          child: Builder(
            builder: widget.builder,
          ),
        ),
      ),
    );
  }
}


lib/src/frame_content.dart
import 'package:flutter/material.dart';

class FrameContent extends StatelessWidget {
  /// Size frame
  final Size size;

  /// Clip behavior
  final Clip clipBehavior;

  /// Child widget
  final Widget child;

  const FrameContent({
    Key? key,
    required this.size,
    required this.clipBehavior,
    required this.child,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return FittedBox(
      child: _screen(context),
    );
  }

  static MediaQueryData mediaQuery(BuildContext context, Size size) {
    final mediaQuery = MediaQuery.of(context);
    final padding = EdgeInsets.symmetric(vertical: 20);

    return mediaQuery.copyWith(
      size: size,
      viewInsets: EdgeInsets.zero,
      viewPadding: padding,
    );
  }

  Widget _screen(BuildContext context) {
    final currentSize = MediaQuery.of(context).size;
    final sizeWidth =
        currentSize.width > size.width ? size.width : currentSize.width;

    final box = SizedBox(
      width: sizeWidth,
      height: MediaQuery.of(context).size.height,
      child: MediaQuery(
        data: FrameContent.mediaQuery(context, size),
        child: child,
      ),
    );

    if (clipBehavior == Clip.none) {
      return box;
    }

    return ClipRect(
      clipBehavior: clipBehavior,
      child: box,
    );
  }
}


lib/src/media_query_observer.dart
import 'package:flutter/widgets.dart';

/// Observes the window changes and provides a [MediaQuery]
/// always up to date.
class MediaQueryObserver extends StatefulWidget {
  final Widget child;

  const MediaQueryObserver({
    required this.child,
  });

  @override
  _MediaQueryObserverState createState() => _MediaQueryObserverState();
}

class _MediaQueryObserverState extends State<MediaQueryObserver>
    with WidgetsBindingObserver {

  T? _ambiguate<T>(T? value) => value;

  @override
  void didChangeMetrics() {
    setState(() {});
    super.didChangeMetrics();
  }

  @override
  void initState() {
    _ambiguate(WidgetsBinding.instance)!.addObserver(this);
    super.initState();
  }

  @override
  void dispose() {
    _ambiguate(WidgetsBinding.instance)!.removeObserver(this);
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return MediaQuery(
      data: MediaQueryData.fromWindow(_ambiguate(WidgetsBinding.instance)!.window),
      child: widget.child,
    );
  }
}


example/lib/main.dart
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:flutter_web_frame/flutter_web_frame.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return FlutterWebFrame(
      builder: (context) {
        return MaterialApp(
          home: FirstScreen(),
        );
      },
      clipBehavior: Clip.hardEdge,
      maximumSize: Size(475.0, 812.0),
      enabled: kIsWeb,
      backgroundColor: Colors.grey,
    );
  }
}

class FirstScreen extends StatelessWidget {
  const FirstScreen({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Flutter Web Frame'),
      ),
      body: Stack(
        children: [
          Positioned(
            child: _buildButton(context),
            left: 0,
            top: 20,
          ),
          Positioned(
            child: _buildButton(context),
            right: 0,
            top: 20,
          ),
          Positioned(
            child: _buildButton(context),
            left: 0,
            bottom: 20,
          ),
          Positioned(
            child: _buildButton(context),
            right: 0,
            bottom: 20,
          ),
          Center(child: _buildButton(context)),
        ],
      ),
    );
  }

  Widget _buildButton(BuildContext context) {
    return ElevatedButton.icon(
      onPressed: () {
        Navigator.of(context).push(
          MaterialPageRoute(
            builder: (_) => SomeScreen(number: 1),
          ),
        );
      },
      icon: Icon(Icons.next_plan),
      label: Text('Next Page'),
    );
  }
}

class SomeScreen extends StatelessWidget {
  const SomeScreen({Key? key, required this.number}) : super(key: key);

  final int number;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Screen $number'),
      ),
      body: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        children: [
          ElevatedButton(
            onPressed: () {
              Navigator.of(context).pop();
            },
            child: Text('Back'),
          ),
          SizedBox(height: 20, width: double.infinity),
          ElevatedButton(
            onPressed: () {
              Navigator.of(context).push(
                MaterialPageRoute(
                  builder: (_) => SomeScreen(number: number + 1),
                ),
              );
            },
            child: Text('Next To Screen ${number + 1}'),
          ),
          SizedBox(height: 20, width: double.infinity),
          TextField(
            decoration: InputDecoration(hintText: 'Input some text'),
          ),
        ],
      ),
    );
  }
}
